File: SCRIPTS/zefania-to-json.py
"""
Convert a Zefania XML Bible file to the flat JSON format used by the COC Bible reader.

Output format:
{
  "1": { "1": { "1": "verse text", "2": "verse text", ... }, ... },
  "2": { ... },
  ...
}

Usage:
  python zefania-to-json.py <input.zefania.xml> <output.json>
"""

import sys
import json
import xml.etree.ElementTree as ET

def convert(src, dst):
    tree = ET.parse(src)
    root = tree.getroot()
    bible = {}

    for book in root.findall('BIBLEBOOK'):
        bnum = book.get('bnumber')
        if not bnum:
            continue
        chapters = {}
        for chapter in book.findall('CHAPTER'):
            cnum = chapter.get('cnumber')
            if not cnum:
                continue
            verses = {}
            for verse in chapter.findall('VERS'):
                vnum = verse.get('vnumber')
                text = (verse.text or '').strip()
                if vnum and text:
                    verses[vnum] = text
            if verses:
                chapters[cnum] = verses
        if chapters:
            bible[bnum] = chapters

    with open(dst, 'w', encoding='utf-8') as f:
        json.dump(bible, f, ensure_ascii=False, separators=(',', ':'))
    print(f"Written {dst}")

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print("Usage: python zefania-to-json.py <input.xml> <output.json>")
        sys.exit(1)
    convert(sys.argv[1], sys.argv[2])

File: SCRIPTS/zefania-to-json.js
// Convert a Zefania XML Bible to the flat JSON format used by the COC Bible reader.
// Usage: node zefania-to-json.js <input.xml> <output.json>

const fs   = require('fs');
const path = require('path');

const [,, src, dst] = process.argv;
if (!src || !dst) {
  console.error('Usage: node zefania-to-json.js <input.xml> <output.json>');
  process.exit(1);
}

const xml   = fs.readFileSync(src, 'utf8');
const bible = {};

const bookRe   = /<BIBLEBOOK\s[^>]*bnumber="(\d+)"[^>]*>([\s\S]*?)<\/BIBLEBOOK>/g;
const chapRe   = /<CHAPTER\s[^>]*cnumber="(\d+)"[^>]*>([\s\S]*?)<\/CHAPTER>/g;
const verseRe  = /<VERS\s[^>]*vnumber="(\d+)"[^>]*>([\s\S]*?)<\/VERS>/g;

let bookMatch;
while ((bookMatch = bookRe.exec(xml)) !== null) {
  const bnum     = bookMatch[1];
  const bookBody = bookMatch[2];
  const chapters = {};

  let chapMatch;
  chapRe.lastIndex = 0;
  while ((chapMatch = chapRe.exec(bookBody)) !== null) {
    const cnum     = chapMatch[1];
    const chapBody = chapMatch[2];
    const verses   = {};

    let vMatch;
    verseRe.lastIndex = 0;
    while ((vMatch = verseRe.exec(chapBody)) !== null) {
      const vnum = vMatch[1];
      const text = vMatch[2]
        .replace(/<[^>]+>/g, '')   // strip any inline tags
        .replace(/&amp;/g,  '&')
        .replace(/&lt;/g,   '<')
        .replace(/&gt;/g,   '>')
        .replace(/&apos;/g, "'")
        .replace(/&quot;/g, '"')
        .trim();
      if (text) verses[vnum] = text;
    }
    if (Object.keys(verses).length) chapters[cnum] = verses;
  }
  if (Object.keys(chapters).length) bible[bnum] = chapters;
}

fs.writeFileSync(dst, JSON.stringify(bible), 'utf8');
console.log(`Written ${dst} — ${Object.keys(bible).length} books`);
